firedrake/mg/ufl_utils.py

```diff
@@ -184,7 +184,7 @@ def coarsen_function(expr, self, coefficient_mapping=None):
     if new is None:
         Vf = expr.function_space()
         Vc = self(Vf, self)
-        new = firedrake.Function(Vc, name=f"coarse_{expr.name()}")
+        new = type(expr)(Vc, name=f"coarse_{expr.name()}")
```

Comment:
This shouldn't be needed as the type is inferred from the function space. A dual space `Vc` would result in a `Cofunction`, while a primal space would produce a `Function`.

Reply:
Ok, I can change it back, I just found it less confusing when I was trying to work out when things switched between primal and dual.